### DISCUSSION TITLE
Firmware Font Customization

### DISCUSSION THREAD

**Discussion by Jyers** [Jyers/Marlin#953] [General Discussion]
Moving this from #869

**Comment by Jyers** [comment]
What better first test that **Comic Sans MS** 😅 
Stick this in your DWIN_SET and let me know if this works
[Comic Sans Font.zip](https://github.com/Jyers/Marlin/files/6587966/Comic.Sans.Font.zip)

**Reply by Cal-LF** [reply]
I see no difference with the font. I use DWIN_SET(Gotcha).

**Reply by Jyers** [reply]
Weird, try this one and see if that works

[Impact Font.zip](https://github.com/Jyers/Marlin/files/6588080/Impact.Font.zip)

**Reply by Cal-LF** [reply]
No difference  with the font. Should I notice the change in all menus?

**Reply by Tuncay-Ayhan** [reply]
@Jyers That is funny 🤣🤣

**Reply by gothcha** [reply]
Didn't change anything for me either, Is there anything that needs to be removed from DWIN?

I tested also with the standard Icon set - nothing!

**Reply by gothcha** [reply]
The font file is 3MB isn't that too large for the display?

**Reply by Jyers** [reply]
It's what the font tool spits out, so it should be correct...

**Reply by adam3654** [reply]
> It's what the font tool spits out, so it should be correct...

Dumb question, but did it work for you?

**Reply by Jyers** [reply]
Unfortunately I don't have time to try it with the other stuff I need to work on

**Reply by adam3654** [reply]
> Unfortunately I don't have time to try it with the other stuff I need to work on

I got as far as putting the dwin folder on my micro SD last night, I'll try it tonight after work. I do remember reading somewhere about a file size limitation. When I get a chance, I'll mess with the font software to see if we can remove unnecessary characters to reduce file size.

I didn't realize that the font file is separate from the bin file, I didn't move on this because I thought the bin file needed to be modified with the contents of the hzk file that the program outputs. 

I really want to see other fonts available, the stock font is visually nauseating to me. I would welcome comic sans in my display over stock. 😉

**Reply by gothcha** [reply]
@adam3654 I thought the same thing, if I remember correctly the file size limit is 512K for the jpg & 9.ICO so the fonts would need to be less than approx 200K

**Reply by gothcha** [reply]
I found this
[T5UIC1.Kernel.Application.Guide.pdf](https://github.com/Jyers/Marlin/files/6593236/T5UIC1.Kernel.Application.Guide.pdf)

**Reply by adam3654** [reply]
@gothcha  I am trying to comprehend the [DWIN development PDF](http://www.ampdisplay.com/documents/pdf/DWIN%20DGUS%20DEV%20GUIDE_V40_2014.pdf) 
and I think the info on page 10 says that the user data is stored in sections 24 through 127 of the flash memory in 256KB (the 26KB shown is assumed to be an error) chunks.  I would think this supports your file size estimate. 

I am still struggling with how the display reads in the files of the DWIN_SET folder. Maybe if I read the PDF a couple dozen more times something might click.

**Reply by adam3654** [reply]
Do you think the font file should be renamed '0T5UIC1.HZK' per the Kernel Application Guide?

**Reply by gothcha** [reply]
hehe yes I am at the same point. It is very confusing how much memory this particular display has. It may be that the file is being loaded but the font that is being called up is still the internal default?

Maybe a change is required to the firmware to address the correct one.

**Reply by gothcha** [reply]
I understand that the leading 0 is the font number and the filename is optional so it shouldn't make any difference

**Reply by Qubi-B** [reply]
Ill test when im home

**Comment by Quint097** [comment]
I've setup a repo with my experimental codes for creating new .HZK Font files: https://github.com/Quint097/DWIN_Font_Files.

I've basically written the reverse of the `decode.go` file provided by @enigmaquip. It translates typeface table pictures back into binary and inserts them into a copy of the existing font file. this way it rewrites only the parts we want and leaves the rest untouched.

**Reply by gothcha** [reply]
@Quint097 Excellent yes I looked through the .go file and guessed that you would likely reverse engineer it. I shall check it out in the morning. 

So who will be the first brave volunteer!

**Reply by enigmaquip** [reply]
Heyo, when old code I forgot about is useful to someone else :-)
Have you guys worked out updating files on the DWIN through marlin instead of having to take apart the screen for the sd slot? It's possible through the serial link to the screen

**Reply by Quint097** [reply]
> Have you guys worked out updating files on the DWIN through marlin instead of having to take apart the screen for the sd slot? It's possible through the serial link to the screen

@enigmaquip Is that how you extracted the original DWIN font file? Because that would mean that it is possible to access the internal display files and that you can make a backup of the files before you start testing.

**Reply by enigmaquip** [reply]
> @enigmaquip Is that how you extracted the original DWIN font file? Because that would mean that it is possible to access the internal display files and that you can make a backup of the files before you start testing.

~~The font file was publicly available on Marlin's configuration github~~ (Hmmm, I found it somewhere lol), I just spent some time iterating through the file to figure out the encoding

**Reply by enigmaquip** [reply]
I was the one who added the documentation for those functions from T5UIC1 documentation I found. It didn't go beyond that because I don't write code in C++. But I could tell you how they're supposed to work.

**Reply by Quint097** [reply]
> But I could tell you how they're supposed to work.

Yes, that would be great.

**Reply by enigmaquip** [reply]
Let's start at the beginning, the T5UIC1 has 16 32kb page files that can hold a full screen jpg, or jpg icon sets that can span multiple pages. We see this is the case with the creality set, boot image on page 0, language images on 1 and 2, and the icon set which comes in at 213kb and spans 6 pages (9-F) [This can be done due to the header at the beginning of the icon file pointing to the location of each icon, so it's always in the first page]. This is also why you can only add a few extra icons before you wont get anymore as it hits the end of the last page.

The max data frame you can send at one time is 255 bytes, we artificially set a smaller list size here
https://github.com/Jyers/Marlin/blob/b4c8ae9b80d896c54733b7580266d7e567c1e2f5/Marlin/src/lcd/dwin/dwin.cpp#L45
This 255 includes the header (AA) and the tail (CC33C33C)

Say we wanted to replace the boot image
We can open our new image off the sdcard in marlin, read it in, and write to the SRAM around 246 bytes at a time, iterating over further and further positions in the memory

https://github.com/Jyers/Marlin/blob/b4c8ae9b80d896c54733b7580266d7e567c1e2f5/Marlin/src/lcd/dwin/dwin.cpp#L450
Here with the Write Data Memory we would
```
size_t i = 0;
DWIN_Byte(i,0x31); //command 0x31
DWIN_Byte(i,0x5A); //write to SRAM
DWIN_Word(i,address to write to); //starting at 0 and obviously increasing each data frame sent
We'd then append our data to the buffer here increasing i the length of the data
DWIN_Send(i);
```

Once we've filled the SRAM with our new image we're gonna write it to a page location (0 in this example as we're overwriting the boot image)

https://github.com/Jyers/Marlin/blob/b4c8ae9b80d896c54733b7580266d7e567c1e2f5/Marlin/src/lcd/dwin/dwin.cpp#L467

```
size_t i = 0;
DWIN_Byte(i, 0x33); //command 0x33
DWIN_Byte(i, 0x5A);
DWIN_Byte(i, 0xA5);
DWIN_Byte(i, PicID); //0x00 for our example
DWIN_Send(i);
```

if we were overwriting the ico set we would have to change pages (picID) every 32kb

This does send a response back after it's finished writing, I don't know if Marlin would handle this (or ignore it, or lock up) as the only function used by marlin now that has a response is in the handshake

https://github.com/Jyers/Marlin/blob/b4c8ae9b80d896c54733b7580266d7e567c1e2f5/Marlin/src/lcd/dwin/dwin.cpp#L92

**Reply by gothcha** [reply]
@enigmaquip This is very useful. If I understand correctly only the SRAM 32KB and the Flash 16KB can be written using the serial port, the internal picture memory and font library can only be loaded using the onboard SD card?

The T5UIC1 kernel documentation does not give the 0x23 DWIN_ICON_Show command that Marlin uses.

Whereas the fonts and icons can be loaded and displayed on the screen from the serial port using the picture reference and the icon file number (9 in our case) I cannot see there is a method of actually installing them into the internal storage via Marlin (or direct serial connection)

So I still believe that the .HZK file is the way to load new fonts. Is this what everyone else thinks?

**Reply by enigmaquip** [reply]
@gothcha you're looking at an older version of the kernel application guide then (I'm guessing v1.2), 0x23 is documented in exactly the way we use it. 0x31, 0x32 and 0x33 are also documented in newer versions of the guide exactly as I have laid out above.  
Yes, I seem to have misremembered the font file as being stored in a page file, so fonts would need to be loaded to the screen in the onboard sdcard method. But jpgs and icons would be as described above.

**Reply by gothcha** [reply]
Yes I have 1.2 - Do you have a link to the newer version?

**Reply by enigmaquip** [reply]
http://www.dwin.com.cn/service/en/file/id/35

**Reply by gothcha** [reply]
Is it supposed to be in Chinese ?

**Reply by enigmaquip** [reply]
I think the last version available in english is 1.2 which is why it's used
as documentation the most

**Reply by Quint097** [reply]
Oke, so this means we can't access the internal font file?

In that case we can only realy just try to upload the new font and see if it works. but if it doesn't, and the original font file we have also isn't correct somehow, then we might end up with a screen that's stuck with a broken font.

Maybe someone how has a (partially) broken screen laying around can try it on that. Then if it breaks the font, it doesn't matter.

**Reply by enigmaquip** [reply]
Quint's NK57 font file
![IMG_20210615_193203](https://user-images.githubusercontent.com/6134440/122145237-190dcc80-ce12-11eb-9635-80cf10314c02.jpg)
Back to my copy of the original font file
![IMG_20210615_193521](https://user-images.githubusercontent.com/6134440/122145244-1b702680-ce12-11eb-8775-b23599b97460.jpg)

**Reply by gothcha** [reply]
Wow you were brave, well done! So we know it works. Now we can find the best font.

**Reply by gothcha** [reply]
Why is only some of the text in the new font? 
maybe it is my imagination?

**Reply by Quint097** [reply]
I think he's using the stock firmware, which is using the eng.jpg for certain parts.

**Reply by gothcha** [reply]
No it is OK, the font is not different enough on all letters to spot the change. Did you test the second font @Quint097 built?

**Reply by gothcha** [reply]
Yes sorry I just realised that it is the old firmware!

**Reply by gothcha** [reply]
I will give it a try, fingers crossed 🤞

**Reply by Sweyn-Forkbeard** [reply]
Huge thumbs up from me on the work you have achieved, and thank you for trying my suggestion of NK57 font, I use it a huge amount because it's easy on my brain, no I don't know how that works, or why, but as I always say, under the skin the human body is a complex and beautiful thing, but so very fragile. I'm also glad it hasn't destroyed my sense of humour and I can still laugh at myself. Top work chaps

**Reply by gothcha** [reply]
@Quint097 What a genius!! It works like a dream. I loaded the Droid font it looks great. Thank you so much for the tools too.

Strangely when I downloaded the zip for your repo it only included the type cards and the illustrator files. I had to select each of the font files individually to download them.

![IMG_5694](https://user-images.githubusercontent.com/80574048/122227515-da7a1f80-ceae-11eb-83ac-fc9db72e96de.jpeg)
![IMG_5695](https://user-images.githubusercontent.com/80574048/122227524-dcdc7980-ceae-11eb-986d-29559671bdc2.jpeg)

**Reply by Quint097** [reply]
Great to hear! I'm glad it's working.

**Reply by enigmaquip** [reply]
@Quint097 I took your encoder a step further. It'll completely automate the process of creating an hzk for any ttf you have in the folder when running it. Still needs some tweaking on positioning and such, but creates a passable hzk.
https://github.com/enigmaquip/DWIN_T5UIC1_Font

**Reply by Quint097** [reply]
The link doesn't work.

**Reply by enigmaquip** [reply]
Fixed, was private

**Reply by Quint097** [reply]
@enigmaquip
I like the idee, but the quality and readability of the characters really diminishes with the smaller sizes and as you already said, the characters are also not centered. Can these things be improved?

**Reply by adam3654** [reply]
I tried installing the fonts from @enogmaquip and got this:

![PXL_20210618_051322488~2.jpg](https://user-images.githubusercontent.com/16920869/122571676-f8649300-d01a-11eb-9a7f-90477dc0899f.jpg)

I had to go back to the original hzk font file provided in that link a while ago, the original font file from the repo did the same thing. I could see that the characters were updating, like the flashing question mark.

Now I'm curious why it works on other devices but not mine!

**Reply by Quint097** [reply]
@adam3654
which exact files did you upload to your display?

**Reply by adam3654** [reply]
@Quint097 I tried all three of the hzk files from the repo folders. I reformatted the card 4k as directed every time. I also tried solo in the DWIN_SET folder, and with the rest of the display firmware files from Jyers repo. I can't really think of anything I could be doing wrong. I use the custom icon set from Jyers repo.

**Reply by Quint097** [reply]
@adam3654
Can you create a issue on the issues tab on my [repo](https://github.com/Quint097/DWIN_Font_Files)? That way i can hopefully help you better.

**Reply by Quint097** [reply]
@adam3654
Might it be that the files just didn't download correctly?
Maybe try these: [HZK Font Files.zip](https://github.com/Jyers/Marlin/files/6677744/HZK.Font.Files.zip)

**Reply by enigmaquip** [reply]
@Quint097 first commit was basically a proof of concept. I've since pushed fixes for the font being undersized and fixes to vertical and horizontal spacing. It's way closer now to a final solution.

**Reply by adam3654** [reply]
> @adam3654
> Might it be that the files just didn't download correctly?
> Maybe try these: [HZK Font Files.zip](https://github.com/Jyers/Marlin/files/6677744/HZK.Font.Files.zip)

@Quint097 The Droid Sans Mono worked from the zip. That's weird because I had downloaded the font files multiple times last night/this morning and they all failed. 

Thank you for the help!

**Comment by Cal-LF** [comment]
That does look better. I would like to try it when you get it sorted out. Maybe post it in a zip.

**Reply by Quint097** [reply]
Here are the font files:
[HZK Font Files.zip](https://github.com/Jyers/Marlin/files/6677431/HZK.Font.Files.zip)

The zip contains:
- DWIN Font (Original)
- DWIN Font (Droid Sans Mono)
- DWIN Font (NK57 Monospace Condensed)
- README.md

**Reply by Cal-LF** [reply]
@Quint097 Thanks for the chance to look at the fonts. I like the Droid Sans Mono better. Can the size be increased, especially the capitals? They both are quite small to read. First is original.
![Image01](https://user-images.githubusercontent.com/76499230/122576808-197bb280-d020-11eb-8ee1-4f91a21409c1.jpg)
![Image02](https://user-images.githubusercontent.com/76499230/122576815-1da7d000-d020-11eb-993b-8a73c9250ee2.jpg)

**Reply by Quint097** [reply]
I'm not sure, I'll see what I can do.

**Reply by Cal-LF** [reply]
@Quint097 Not criticizing, just seeing readability is better with larger capitals.

**Reply by Quint097** [reply]
I have created this one: 
[DWIN Font (Droid Sans Mono - Resized).zip](https://github.com/Jyers/Marlin/files/6678481/DWIN.Font.Droid.Sans.Mono.-.Resized.zip)

I think it might be a bit oversized now, but let me know what you think.

**Reply by Cal-LF** [reply]
@Quint097 This is what I am seeing. Much better.

![Image01](https://user-images.githubusercontent.com/76499230/122601616-90747380-d03f-11eb-8ad4-0b13d4f1994d.jpg)
![Image01](https://user-images.githubusercontent.com/76499230/122603996-41c8d880-d043-11eb-82b5-faf75c7b260d.jpg)

**Reply by gothcha** [reply]
Yes @Quint097 I preferred the Droid too, and the larger capitals do look great, I shall install the new version tomorrow. Thanks for all the hard work.

**Reply by Cal-LF** [reply]
@Quint097 @gothcha Yes, keeping this installed. I like it.

**Reply by gothcha** [reply]
@Quint097 Loaded the enlarged version, looks good :). If I am being picky the 0 is now not symmetrical it looks like it is shifted so the vertical left and right edges are not the same number of pixel tall, ty is visible in the images above if you look closely.

**Reply by Cal-LF** [reply]
I see the issue with what looks like unequal or non symmetrical sides to the zeros. Whereas the not resized is equal.

**Reply by Quint097** [reply]
@Cal-LF @gothcha
I have fixed the zero problem. You can download the most recent version of the fonts form the release tab on my repo.

**Reply by Quint097** [reply]
@tititopher68-dev
I'm not sure I'm understanding what you're saying.

**Reply by Cal-LF** [reply]
@Quint097 The date on the files is from yesterday.

**Reply by Quint097** [reply]
@Cal-LF
Yes, that's when i fixed it.

**Reply by Cal-LF** [reply]
@Quint097 Flashed successfully and it looks nice. Your hard work on this font project has made the display easier to read. I do appreciate this improvement.

**Reply by Quint097** [reply]
> @Quint097 Flashed successfully and it looks nice. Your hard work on this font project has made the display easier to read. I do appreciate this improvement.

@Cal-LF Thanks!

By the way; I'm sorry if I'm sometimes a bit blunt with my answers, I'm not the best at communicating via text.

**Reply by Cal-LF** [reply]
LOL. We aren't writing resumes or love letters.

**Comment by Quint097** [comment]
If it's alright with everyone, I would like to move any font related issues, questions and discussions to [DWIN_Font_Tools](https://github.com/Quint097/DWIN_Font_Tools) to keep things organised.

I have created a discussion threads on the new release, so any questions or issues related to that can be posted there.

**Reply by Cal-LF** [reply]
I will follow from your Github.

**Reply by Jyers** [reply]
@Quint097 I'll link that around where appropriate 👌
